src/wal/record.rs

```diff
@@ -89,72 +99,45 @@ impl Record {
         let mut raw_record = [0u8; WAL_RECORD_SIZE];
         raw_record[0..8].copy_from_slice(&self.sequence_number.to_le_bytes());
         // skip checksum, fill later
+
         let entry_count: u16 = self
             .data
             .len()
             .try_into()
             .expect("Entries must fit in WAL_RECORD_SIZE");
 
-        let mut offset = 14;
+        let last_indicator = if self.last {
+            LAST_IN_SEQUENCE
+        } else {
+            NOT_LAST_IN_SEQUENCE
+        };
+
+        raw_record[14] = last_indicator;
+
+        let mut offset = DATA_START;
         for entry in self.data {
             let entry_len = entry.len();
             raw_record[offset..offset + entry_len].copy_from_slice(&entry.to_bytes());
             offset += entry_len;
         }
 
-        let checksum: u32 = {
-            raw_record[10..14].copy_from_slice(&WAL_CHECKSUM_MAGIC.to_le_bytes());
-            let mut buf = [0u8; 4];
-            let checksum_hash = blake3::hash(&raw_record[10..offset]);
-            buf.copy_from_slice(&checksum_hash.as_bytes()[0..4]);
-            u32::from_le_bytes(buf)
-        };
+        let checksum = checksum(last_indicator, &raw_record[DATA_START..]);
 
         raw_record[8..12].copy_from_slice(&checksum.to_le_bytes());
         raw_record[12..14].copy_from_slice(&entry_count.to_le_bytes());
         raw_record
     }
 }
 
-#[cfg(test)]
-pub mod tests {
-    use super::Record;
-    use crate::wal::entry::Entry;
-
-    pub fn create_record(sequence_number: u64, range: std::ops::Range<u64>) -> Record {
-        let make_entry = |i: u64| -> Entry {
-            if i % 2 == 0 {
-                Entry::Clear { bucket_index: i }
-            } else {
-                Entry::Update {
-                    page_id: {
-                        let mut page_id = [0; 16];
-                        page_id[0..8].copy_from_slice(&i.to_le_bytes());
-                        page_id
-                    },
-                    page_diff: {
-                        let mut page_diff = [0; 16];
-                        page_diff[8..16].copy_from_slice(&i.to_le_bytes());
-                        page_diff
-                    },
-                    changed: (0..i.count_ones()).map(|i| [i as u8; 32]).collect(),
-                    bucket_index: i,
-                }
-            }
-        };
+fn checksum(last: u8, data: &[u8]) -> u32 {
+    let mut buf = [0u8; 4];
+    let mut hasher = blake3::Hasher::new();
+    hasher.update(&WAL_CHECKSUM_MAGIC.to_le_bytes());
```

Comment:
This usage of the hasher makes the checksum creation way cooler than what I previously did!

Reply:
Thanks!